### DISCUSSION TITLE
Export feature

### DISCUSSION THREAD

**Discussion by clawes** [filamentphp/filament#471] [Ideas]
The ability to export listings to Excel, CSV or PDF file would be nice to have.

**Comment by anishsarkar** [comment]
After struggling for couple of hours, I found that laravel-excel can be used for exporting. I use only CSV, you may modify these as per you needs. 

Here are my codes

## Resource UserResource 

### ListUsers

```

use Maatwebsite\Excel\Excel; 

...

class ListUsers extends ListRecords
{

    protected static string $resource = UserResource::class;

    protected function getTableBulkActions(): array
    {
        return [
            BulkAction::make('export')
                ->action( fn (Collection $records) => new UsersExport($records))
                ->deselectRecordsAfterCompletion()
        ];
    }
}

```

### UsersExport

```

use Maatwebsite\Excel\Excel;
use Illuminate\Contracts\Support\Responsable;
use Illuminate\Support\Collection;
use Maatwebsite\Excel\Concerns\FromCollection;
use Maatwebsite\Excel\Concerns\Exportable;

class UsersExport implements FromCollection, Responsable
{
    use Exportable;


    private Collection $records;
    
    private $fileName = 'users.csv';
    
    private $writerType = Excel::CSV;
    
    private $headers = [
        'Content-Type' => 'text/csv',
    ];

    public function __construct(Collection $records)
    {
        $this->records = $records;
    }

    public function collection()
    {
        return $this->records;
    }
}

```
These seems to work. 

One comment:  it would have better if we could get the id of the selected records. Then those could be passed to the **UsersExport** class and instead of a collection, a query method may be
used. The query can be easily queued and that way it would not be a problem to export large datasets.

**Comment by zdnk** [comment]
Since there was native import action introduced recently, it would be awesome if export action is also natively supported.

I am willing to contribute if @zepfietje, @danharrin  or someone else gives me pointers on how to implement this action.

My thoughts:
- use currently displayed columns in the table
- let user choose if it should export current page only or all the data
- needs to be working for relations as well (relation managers)

I guess what I am looking for is:
- how do I create the action? regular action? table action?
- how do I get the query for currently displayed columns and applied filters for a) current page b) selected records c) all records

**Reply by danharrin** [reply]
Hi @zdnk! Thanks for the offer, I have created issue #10117.

I think I'm going to take this on myself for v3.2, but if you want to do some investigation into the best way to performantly export XLSX using batched jobs, that would be great! I know how to do it with CSVs already, but I have a feeling that it might be more difficult with XLS. Any investigation you do will help me when I come to build the export feature 👍

I do want to avoid using the Laravel Excel package as it is very heavy, and we can make use of a simpler native PHP XLS package, alongside `league/csv`.

**Reply by zdnk** [reply]
Well I need this feature next week ready, so I am considering doing the CSV myself as I dont think you guys will be releasing 3.2 before the end of the next week? :) 

Thats why I was asking for pointers and information mentioned above. I happy to contribute back and even pledge some recurring donations from my company to support further development as we love filament so far and are gonna use it with more of our clients and projects.

**Reply by danharrin** [reply]
Why not use this plugin in the meantime? https://filamentphp.com/plugins/pxlrbt-excel

**Reply by zdnk** [reply]
We can do that definitely, just wanted us to contribute. We will use the plugin in the meantime then.

**Reply by danharrin** [reply]
I do appreciate the offer, and if you want to help me I have described how

**Reply by zdnk** [reply]
I tried the mentioned plugin. We have over 30k records in the table I'd like to export and the export times out at 30s. It would be good to somehow consider these limitations as well. Maybe do some sort of background/queue processing of these and make them available later when finished?

**Reply by danharrin** [reply]
Yeah, that's what I want the investigation about. I'm sure I know how to process CSVs in the background, but not Excel, if you need that.

**Comment by cawecoy** [comment]
Hi community. Filament 3.2 introduced the [export action](https://filamentphp.com/docs/3.x/actions/prebuilt-actions/export) but there is no option for PDF file type. How can I extend the export action in order to allow the content to be exported and downloaded in PDF file type? 

Maybe I can use [this plugin](https://filamentphp.com/plugins/pxlrbt-excel) which supports PDF filetype, but will it become obsolete with the Filament pre-build export action?
